### PR TITLE
Testing Round One

### PR DESCRIPTION
**Pull request by DavidVorick**
Added some stuff, including:

1. `make cover`, which produces testing coverage stats for all of our packages that have test files. Eventually all packages should have test files.

2. documentation to the makefile

3. Some organizational stuff

4. A whole bunch of tests for the consensus package, bringing coverage up to 44%. Right now pretty much only empty blocks are tested, which is why the number is so low. No contracts or financial stuff is tested, and the exported functions don't get much love either. That'll be in round two.

### CODE REVIEW

**Comment by DavidVorick** [comment]
Okay added a bunch more stuff.

Some more tests, some changes to how block ids are calculated, and some extra organizational things.

File: consensus/blocks_test.go [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+	// Compare the metrics and report an error if something has changed.
+	if bbLen != len(s.badBlocks) ||
+		bmLen != len(s.blockMap) ||
+		mpLen != len(s.missingParents) ||
+		cpLen != len(s.currentPath) ||
+		uoLen != len(s.unspentOutputs) ||
+		ocLen != len(s.openContracts) ||
+		stateHash != s.StateHash() {
+		t.Error("state changed after getting a repeat block.")
+	}
```
Comment (lukechampine) [inline consensus/blocks_test.go]: Isn't it sufficient just to recalculate and compare the state hash?

File: consensus/blocks_test.go [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+// TestBlockTimestamps creates a new state and uses it to call
+// testBlockTimestamps.
+func TestBlockTimestamps(t *testing.T) {
+	s := CreateGenesisState()
+	testBlockTimestamps(t, s)
+}
+
+// TestConstants makes sure that the testing constants are being used instead
+// of the developer constants or the release constants.
```
Comment (lukechampine) [inline consensus/blocks_test.go]: this is getting a little ridiculous

File: consensus/types.go [hunk 3]
```diff
@@ -135,12 +132,14 @@ func (b Block) CheckTarget(target Target) bool {
...
 // ExpectedTransactionMerkleRoot() returns the expected transaction
 // merkle root of the block.
-func (b Block) TransactionMerkleRoot() hash.Hash {
-	var transactionHashes []hash.Hash
-	for _, transaction := range b.Transactions {
-		transactionHashes = append(transactionHashes, hash.HashBytes(encoding.Marshal(transaction)))
+func (b Block) MerkleRoot() hash.Hash {
+	var hashes []hash.Hash
+	hashes = append(hashes, hash.HashObject(b.Timestamp))
+	hashes = append(hashes, hash.HashObject(b.MinerAddress))
```
Comment (lukechampine) [inline consensus/types.go]:
```go
hashes := []hash.Hash{
    hash.HashObject(b.Timestamp)),
    hash.HashObject(b.MinerAddress)),
}
```

Comment (DavidVorick) [inline consensus/types.go]: Not worried about performance right now, just something that's clean and easy to read.

I'd imagine that consecutive appends get optimized by the compiler anyway.

File: consensus/blocks_test.go (hunk 2 above)
Comment (DavidVorick) [inline consensus/blocks_test.go]: what is?

**Pull request merged**